sql_execution_engine/utils.go

```diff
@@ -0,0 +1,175 @@
+package sql_execution_engine
+
+import (
+	"database/sql"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/juju/errors"
+	"github.com/moiot/gravity/gravity/registry"
+	"github.com/moiot/gravity/pkg/core"
+	"github.com/moiot/gravity/schema_store"
+	"github.com/sirupsen/logrus"
+)
+
+func GenerateSingleDeleteSQL(msg *core.Msg, tableDef *schema_store.Table) (string, []interface{}, error) {
+	primaryKeyData := msg.DmlMsg.Pks
+	var whereSql []string
+
+	var args []interface{}
+	for _, column := range tableDef.Columns {
+		columnName := column.Name
+		pkData, ok := primaryKeyData[columnName]
+		if !ok {
+			continue
+		}
+
+		whereSql = append(whereSql, fmt.Sprintf("`%s` = ?", columnName))
+		args = append(args, adjustArgs(pkData, tableDef.MustColumn(columnName)))
+	}
+	if len(whereSql) == 0 {
+		return "", nil, errors.Errorf("where sql is empty, probably missing pk")
+	}
+
+	stmt := fmt.Sprintf("DELETE FROM `%s`.`%s` WHERE %s", tableDef.Schema, tableDef.Name, strings.Join(whereSql, " AND "))
+	return stmt, args, nil
+}
+
+func GenerateReplaceSQLWithMultipleValues(msgBatch []*core.Msg, tableDef *schema_store.Table) (string, []interface{}, error) {
+	columnNames := make([]string, len(tableDef.Columns))
+	for _, column := range tableDef.Columns {
+		columnName := column.Name
+		columnIdx := column.Idx
+		columnNames[columnIdx] = fmt.Sprintf("`%s`", columnName)
+	}
+
+	sqlPrefix := fmt.Sprintf("REPLACE INTO `%s`.`%s` (%s) VALUES", tableDef.Schema, tableDef.Name, strings.Join(columnNames, ","))
+
+	// Generate place holders and args
+	batchPlaceHolders, args, err := PlaceHoldersAndArgsFromEncodedData(msgBatch, tableDef)
+	if err != nil {
+		return "", nil, errors.Trace(err)
+	}
+
+	finalPlaceHolders := strings.Join(batchPlaceHolders, ",")
+	s := []string{sqlPrefix, finalPlaceHolders}
+	return strings.Join(s, " "), args, nil
+}
+
+func PlaceHoldersAndArgsFromEncodedData(msgBatch []*core.Msg, tableDef *schema_store.Table) ([]string, []interface{}, error) {
+	var batchPlaceHolders []string
+	var batchArgs []interface{}
+
+	for _, msg := range msgBatch {
+		if err := ValidateSchema(msg, tableDef); err != nil {
```

Comment:
how should we handle the case target table has more column than source? As we mentioned the `Default` keyword before.

Reply:
deal with this case later